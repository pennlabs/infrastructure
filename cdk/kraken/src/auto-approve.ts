import { Workflow, Stack, WorkflowProps, Job } from "cdkactions";
import { Construct } from "constructs";

/**
 * Auto-approve PRs opened by dependabot. Used to fulfil requirements
 * needed to automatically merge using mergify
 */
export class AutoApproveStack extends Stack {
  /**
   *
   * @param scope cdkactions App instance.
   * @param overrides Optional overrides for the stack.
   */
  public constructor(scope: Construct, overrides?: Partial<WorkflowProps>) {
    super(scope, "autoapprove");
    const workflow = new Workflow(this, "autoapprove", {
      name: "Auto Approve dependabot PRs",
      on: "pullRequest",
      ...overrides,
    });

    new Job(workflow, "approve", {
      runsOn: "ubuntu-latest",
      steps: [
        {
          uses: "hmarr/auto-approve-action@v2.0.0",
          if: "github.actor == 'dependabot[bot]'",
          with: {
            "github-token": "${{ secrets.BOT_GITHUB_PAT }}",
          },
        },
      ],
    });
  }
}
