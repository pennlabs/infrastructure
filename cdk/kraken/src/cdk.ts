import { CheckoutJob, Workflow, Stack, WorkflowProps } from 'cdkactions';
import { Construct } from 'constructs';
import dedent from 'ts-dedent';

/**
 * Optional props to configure the CDK publish stack.
 */
export interface CDKPublishStackProps {
  /**
   * Branch to restrict publishing to
   * @default master
   */
  defaultBranch?: string;

  /**
   * Node version to test the cdk application in
   * @default "14"
   */
  nodeVersion?: string;
}

/**
 * Test, upload code coverage for, and publish a cdk application.
 */
export class CDKPublishStack extends Stack {
  /**
   *
   * @param scope cdkactions App instance.
   * @param id Id of the job.
   * @param config Optional configuration for the cdk publish stack.
   * @param overrides Optional overrides for the stack.
   */
  public constructor(scope: Construct, id: string, config?: CDKPublishStackProps, overrides?: Partial<WorkflowProps>) {
    // Build config
    const fullConfig: Required<CDKPublishStackProps> = {
      defaultBranch: 'master',
      nodeVersion: '14',
      ...config,
    };
    const path = `cdk/${id}`;
    super(scope, id);
    const workflow = new Workflow(this, id, {
      name: `Publish ${id}`,
      on: {
        push: {
          paths: [`${path}/**`],
        },
      },
      ...overrides,
    });
    new CheckoutJob(workflow, 'publish', {
      runsOn: 'ubuntu-latest',
      steps: [{
        name: 'Cache',
        uses: 'actions/cache@v2',
        with: {
          path: '**/node_modules',
          key: `v0-\${{ hashFiles('${path}/yarn.lock') }}`,
        },
      },
      {
        name: 'Install Dependencies',
        run: dedent`cd ${path}
        yarn install --frozen-lockfile`,
      },
      {
        name: 'Test',
        run: dedent`cd ${path}
        yarn test`,
      },
      {
        name: 'Upload Code Coverage',
        run: dedent`ROOT=$(pwd)
        cd ${path}
        yarn run codecov -p $ROOT -F ${id}`,
      },
      {
        name: 'Install jq',
        run: dedent`
        curl -sSLo /usr/bin/jq https://github.com/stedolan/jq/releases/download/jq-1.6/jq-linux64
        chmod +x /usr/bin/jq`,
      },
      {
        name: 'Check if version is already published to npm',
        run: dedent`cd ${path}
        PACKAGE=\$(node -p "require('./package.json').name")
        VERSION=\$(node -p "require('./package.json').version")
        NEW_VERSION=\$(yarn info $PACKAGE versions --json | jq ".data | any(. == \\"$VERSION\\") | not")
        echo "::set-output name=NEW_VERSION::$NEW_VERSION"`,
      },
      {
        name: 'Publish to npm',
        id: 'version_check',
        run: dedent`cd ${path}
        yarn compile
        yarn package
        mv dist/js/*.tgz dist/js/${id}.tgz
        yarn publish --non-interactive --access public dist/js/${id}.tgz`,
        if: `github.ref == 'refs/heads/${fullConfig.defaultBranch}' && steps.version_check.outputs.NEW_VERSION`,
        env: {
          NPM_AUTH_TOKEN: '${{ secrets.NPM_AUTH_TOKEN }}',
        },
      },
      {
        name: 'Build docs',
        run: dedent`cd ${path}
        yarn docgen`,
      },
      {
        name: 'Publish docs',
        if: `github.ref == 'refs/heads/${fullConfig.defaultBranch}'`,
        uses: 'peaceiris/actions-gh-pages@v3',
        with: {
          personal_token: '${{ secrets.BOT_GITHUB_PAT }}',
          external_repository: `pennlabs/${id}-docs`,
          cname: `${id}.pennlabs.org`,
          publish_branch: 'master',
          publish_dir: `${path}/docs`,
          user_name: 'github-actions',
          user_email: 'github-actions[bot]@users.noreply.github.com',
        },
      }],
      container: {
        image: `node:${fullConfig.nodeVersion}`,
      },
    });
  }
}
