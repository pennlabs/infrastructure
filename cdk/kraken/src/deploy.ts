import { CheckoutJob, Workflow, JobProps } from "cdkactions";
import dedent from "ts-dedent";

/**
 * Optional props to configure the deploy job.
 */
export interface DeployJobProps {
  /**
   * Deploy tag to set in kittyhawk.
   * @default current git sha
   */
  deployTag?: string;

  /**
   * Branch to limit deploys to.
   * @default master
   */
  defaultBranch?: string;
}

/**
 * A job to deploy an application using Kittyhawk.
 */
export class DeployJob extends CheckoutJob {
  /**
   *
   * @param scope cdkactions Workflow instance
   * @param config Optional configuration for the deploy job.
   * @param overrides Optional overrides for the job.
   */
  public constructor(
    scope: Workflow,
    config?: DeployJobProps,
    overrides?: Partial<JobProps>
  ) {
    // Build config
    const fullConfig: Required<DeployJobProps> = {
      deployTag: "${{ github.sha }}",
      defaultBranch: "master",
      ...config,
    };

    super(scope, "deploy", {
      runsOn: "ubuntu-latest",
      if: `github.ref == 'refs/heads/${fullConfig.defaultBranch}'`,
      steps: [
        {
          name: 'Checkout kube-manifests',
          uses: 'actions/checkout@v2',
          with: {
            repository: 'pennlabs/kube-manifests',
            token: '${{ secrets.BOT_GITHUB_PAT }}',
            path: 'kube-manifests',
          }
        },
        {
          name: 'Configure git',
          run: dedent`git config --global user.name github-actions
          git config --global user.email github-actions[bot]@users.noreply.github.com"`
        },
        {
          id: 'synth',
          name: 'Synth cdk8s manifests',
          run: dedent`cd k8s
          yarn install --frozen-lockfile

          # get repo name (by removing owner/organization)
          export RELEASE_NAME=\${REPOSITORY#*/}

          # Export RELEASE_NAME as an output
          echo "::set-output name=RELEASE_NAME::$RELEASE_NAME"

          yarn build`,
          env: {
            GIT_SHA: fullConfig.deployTag,
            REPOSITORY: "${{ github.repository }}",
            AWS_ACCOUNT_ID: "${{ secrets.AWS_ACCOUNT_ID }}",
          },
        },
        {
          name: 'Push to kube-manifests repository',
          run: dedent`cd kube-manifests
          mkdir -p \${{ github.repository }}
          cp -r ../k8s/dist/ \${{ github.repository }}

          # get repo name from synth step
          RELEASE_NAME=\${{ steps.synth.outputs.RELEASE_NAME }}

          git add \${{ github.repository }}
          git commit -m "chore(k8s): deploy $RELEASE_NAME"
          git push`
        },
        {
          name: "Deploy",
          run: dedent`aws eks --region us-east-1 update-kubeconfig --name production --role-arn arn:aws:iam::\${AWS_ACCOUNT_ID}:role/kubectl
          # get repo name from synth step
          RELEASE_NAME=\${{ steps.synth.outputs.RELEASE_NAME }}
          # Deploy
          kubectl apply -f k8s/dist/ -l app.kubernetes.io/component=certificate
          kubectl apply -f k8s/dist/ --prune -l app.kubernetes.io/part-of=$RELEASE_NAME`,
          env: {
            AWS_ACCOUNT_ID: "${{ secrets.AWS_ACCOUNT_ID }}",
            AWS_ACCESS_KEY_ID: "${{ secrets.GH_AWS_ACCESS_KEY_ID }}",
            AWS_SECRET_ACCESS_KEY: "${{ secrets.GH_AWS_SECRET_ACCESS_KEY }}",
          },
        },
      ],
      ...overrides,
    });
  }
}
