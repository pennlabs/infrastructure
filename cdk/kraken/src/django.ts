import { CheckoutJob, Workflow, StepsProps, JobProps } from "cdkactions";
import dedent from "ts-dedent";
import { buildId, buildName } from "./utils";

/**
 * Props to configure the Django check job.
 */
export interface DjangoCheckJobProps {
  /**
   * A custom id to append onto job name and ids. Useful when using
   * multiple instances of DjangoCheckJob in a single workflow.
   * @default no suffix
   */
  id?: string;

  /**
   * Python version to test the project with.
   * @default "3.10-bookworm"
   */
  pythonVersion?: string;

  /**
   * Location of the Django project within the repo.
   * @default "."
   */
  path?: string;

  /**
   * Name of the Django project to test.
   */
  projectName: string;

  /**
   * Check the project with black?
   * @default true
   */
  black?: boolean;

  /**
   * Check the project with flake8?
   * @default true
   */
  flake8?: boolean;
}

/**
 * A job to lint and test a Django project as well as upload code coverage.
 *
 * Note that tests are run in parallel to reduce testing time.
 */
export class DjangoCheckJob extends CheckoutJob {
  /**
   *
   * @param scope cdkactions Workflow instance.
   * @param config Configuration for the Django check job.
   * @param overrides Optional overrides for the job.
   */
  public constructor(
    scope: Workflow,
    config: DjangoCheckJobProps,
    overrides?: Partial<JobProps>
  ) {
    // Build config
    const fullConfig: Required<DjangoCheckJobProps> = {
      id: "",
      pythonVersion: "3.10-bookworm",
      path: ".",
      black: true,
      flake8: true,
      ...config,
    };

    // Define steps
    const steps: StepsProps[] = [
      {
        name: "Cache",
        uses: "actions/cache@v4",
        with: {
          path: "~/.local/share/virtualenvs",
          key: `v0-\${{ hashFiles('${fullConfig.path}/Pipfile.lock') }}`,
        },
      },
      {
        name: "Install Dependencies",
        run: dedent`cd ${fullConfig.path}
      pip install pipenv
      pipenv install --deploy --dev`,
      },
    ];
    if (fullConfig.flake8) {
      steps.push({
        name: "Lint (flake8)",
        run: dedent`cd ${fullConfig.path}
        pipenv run flake8 .`,
      });
    }
    if (fullConfig.black) {
      steps.push({
        name: "Lint (black)",
        run: dedent`cd ${fullConfig.path}
        pipenv run black --check .`,
      });
    }
    steps.push({
      name: "Test (run in parallel)",
      run: dedent`cd ${fullConfig.path}
      pipenv run coverage run --concurrency=multiprocessing manage.py test --settings=${fullConfig.projectName}.settings.ci --parallel
      pipenv run coverage combine
      pipenv run coverage xml`,
    });
    steps.push({
      name: "Upload Code Coverage",
      uses: "codecov/codecov-action@v3",
      with: {
        token: "${{ secrets.CODECOV_TOKEN }}",
        directory: "./backend/",
        fail_ci_if_error: true,
        files: "coverage.xml",
        name: "codecov-umbrella",
        verbose: true,
      },
    });

    // Create Job
    super(scope, buildId("django-check", fullConfig.id), {
      name: buildName("Django Check", fullConfig.id),
      runsOn: "ubuntu-latest",
      steps,
      container: {
        image: `python:${fullConfig.pythonVersion}`,
      },
      env: {
        DATABASE_URL: "postgres://postgres:postgres@postgres:5432/postgres",
      },
      services: {
        postgres: {
          image: "postgres:12",
          env: {
            POSTGRES_USER: "postgres",
            POSTGRES_DB: "postgres",
            POSTGRES_PASSWORD: "postgres",
          },
          options:
            "--health-cmd pg_isready --health-interval 10s --health-timeout 5s --health-retries 5",
        },
      },
      ...overrides,
    });
  }
}
