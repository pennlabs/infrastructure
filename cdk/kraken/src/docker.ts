import { CheckoutJob, Workflow, JobProps, StepsProps } from "cdkactions";
import { buildId, buildName } from "./utils";

/**
 * Props to configure the Docker publish job.
 */
export interface DockerPublishJobProps {
  /**
   * Image name to publish
   */
  imageName: string;

  /**
   * Path to the docker context
   * @default "."
   */
  path?: string;

  /**
   * Path to the dockerfile (relative to `path`)
   * @default "Dockerfile"
   */
  dockerfile?: string;

  /**
   * A condition on when to push the built image.
   * @default Only push on master.
   */
  push?: string;

  /**
   * Tags to apply to the built image.
   * @default latest and the current git sha.
   */
  tags?: string;

  /**
   * Username to authenticate with Docker Hub.
   * @default load the "DOCKER_USERNAME" GitHub Actions secret.
   */
  username?: string;

  /**
   * Password to authenticate with Docker Hub.
   * @default load the "DOCKER_PASSWORD" GitHub Actions secret.
   */
  password?: string;

  /**
   * If enabled, will cache_from the latest version of the docker image.
   * @default true
   */
  cache?: boolean;

  /**
   * If evaluates to true, do not publish docker image to Docker Hub.
   * Instead upload as an artifact.
   * @default false
   */
  noPublish?: boolean | string;
}

/**
 * A job to build and publish a Docker image.
 */
export class DockerPublishJob extends CheckoutJob {
  /**
   * Full name of the docker image built.
   */
  public readonly dockerImageName: string;

  /**
   *
   * @param scope cdkactions Workflow instance.
   * @param id Id of the job.
   * @param config Configuration for the docker publish job.
   * @param overrides Optional overrides for the job.
   */
  public constructor(
    scope: Workflow,
    id: string,
    config: DockerPublishJobProps,
    overrides?: Partial<JobProps>
  ) {
    // Build config
    const fullConfig: Required<DockerPublishJobProps> = {
      path: ".",
      push: "${{ github.ref == 'refs/heads/master' }}",
      tags: "latest,${{ github.sha }}",
      username: "${{ secrets.DOCKER_USERNAME }}",
      password: "${{ secrets.DOCKER_PASSWORD }}",
      dockerfile: "Dockerfile",
      cache: true,
      noPublish: false,
      ...config,
    };
    const formattedName = fullConfig.noPublish
      ? buildName("Build", id)
      : buildName("Publish", id);
    const formattedId = fullConfig.noPublish
      ? buildId("build", id)
      : buildId("publish", id);
    // Define docker action with
    const imageName = `pennlabs/${fullConfig.imageName}`;
    const dockerWith: any = {
      context: fullConfig.path,
      file: `${fullConfig.path}/${fullConfig.dockerfile}`,
      push: fullConfig.push,
      "cache-from": "type=local,src=/tmp/.buildx-cache",
      "cache-to": "type=local,dest=/tmp/.buildx-cache",
    };
    if (fullConfig.cache) {
      dockerWith["cache-from"] = dockerWith["cache-from"].concat(
        `,type=registry,ref=${imageName}:latest`
      );
    }

    // Build tags string
    const tagsString = (tags: string) =>
      tags
        .split(",")
        .map((tag) => `${imageName}:${tag}`)
        .join();
    dockerWith.tags = tagsString(fullConfig.tags);

    // Define steps
    const steps: StepsProps[] = [
      {
        uses: "docker/setup-qemu-action@v1",
      },
      {
        uses: "docker/setup-buildx-action@v1",
      },
      {
        name: "Cache Docker layers",
        uses: "actions/cache@v2",
        with: {
          path: "/tmp/.buildx-cache",
          key: `buildx-${formattedId}`,
        },
      },
      {
        uses: "docker/login-action@v1",
        with: {
          username: fullConfig.username,
          password: fullConfig.password,
        },
      },
      {
        name: "Build/Publish",
        uses: "docker/build-push-action@v2",
        with: dockerWith,
      },
    ];

    // If publishing isn't wanted
    if (fullConfig.noPublish) {
      dockerWith.push = false;
      dockerWith.outputs = "type=docker,dest=/tmp/image.tar";
      steps.push({
        uses: "actions/upload-artifact@v2",
        with: {
          name: formattedId,
          path: "/tmp/image.tar",
        },
      });
      steps.splice(3, 1);
    }

    // Create job
    super(scope, formattedId, {
      name: formattedName,
      runsOn: "ubuntu-latest",
      steps,
      ...overrides,
    });

    // Set public fields
    this.dockerImageName = imageName;
  }
}
