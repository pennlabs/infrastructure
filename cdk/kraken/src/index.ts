export * from './cdk';
export * from './deploy';
export * from './django-project';
export * from './django';
export * from './docker';
export * from './integration-tests';
export * from './labs-application';
export * from './pypi';
export * from './react-project';
export * from './react';
export * from './utils';
