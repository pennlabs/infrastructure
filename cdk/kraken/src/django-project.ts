import { JobProps, Workflow } from "cdkactions";
import { DjangoCheckJob, DjangoCheckJobProps } from "./django";
import { DockerPublishJob, DockerPublishJobProps } from "./docker";
import { buildId } from "./utils";

/**
 * Props to configure the DjangoProject.
 */
export interface DjangoProjectProps {
  /**
   * A custom id to append onto job name and ids. Useful when using
   * multiple instances of DjangoProject in a single workflow.
   * @default no suffix
   */
  id?: string;

  /**
   * Name of the Django project to test.
   */
  projectName: string;

  /**
   * Location of the Django project within the repo.
   * @default "."
   */
  path?: string;

  /**
   * Docker image name to publish
   */
  imageName: string;

  /**
   * Optional props to pass to the check job.
   */
  checkProps?: Partial<DjangoCheckJobProps>;

  /**
   * Optional overrides for the check job.
   */
  checkOverrides?: Partial<JobProps>;

  /**
   * Optional props to pass to the docker publish job.
   */
  publishProps?: Partial<DockerPublishJobProps>;

  /**
   * Optional overrides for the docker publish job.
   */
  publishOverrides?: Partial<JobProps>;
}

/**
 * Creates a DjangoCheckJob and a DockerPublishJob to lint, test, build, and publish
 * a docker image of a django project. The jobs run sequentially.
 */
export class DjangoProject {
  /**
   * ID of the docker publish job.
   */
  public readonly publishJobId: string;

  /**
   * Full name of the docker image built.
   */
  public readonly dockerImageName: string;

  /**
   *
   * @param workflow cdkaction Workflow instance.
   * @param config Configuration for the django project,
   */
  public constructor(workflow: Workflow, config: DjangoProjectProps) {
    // Build config
    const fullConfig: Required<DjangoProjectProps> = {
      id: "",
      path: ".",
      checkProps: {},
      checkOverrides: {},
      publishProps: {},
      publishOverrides: {},
      ...config,
    };

    // Add jobs
    const djangoCheckJob = new DjangoCheckJob(
      workflow,
      {
        id: fullConfig.id,
        projectName: fullConfig.projectName,
        path: fullConfig.path,
        ...config.checkProps,
      },
      fullConfig.checkOverrides
    );

    const publishJob = new DockerPublishJob(
      workflow,
      buildId("backend", fullConfig.id),
      {
        imageName: fullConfig.imageName,
        path: fullConfig.path,
        ...fullConfig.publishProps,
      },
      {
        needs: djangoCheckJob.id,
        ...fullConfig.publishOverrides,
      }
    );

    // Set public fields
    this.publishJobId = publishJob.id;
    this.dockerImageName = publishJob.dockerImageName;
  }
}
