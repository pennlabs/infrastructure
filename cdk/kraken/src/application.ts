import { Workflow, JobProps, WorkflowProps, Stack, CheckoutJob } from 'cdkactions';
import { Construct } from 'constructs';
import { DeployJob, DeployJobProps } from './deploy';
import { DjangoCheckJobProps } from './django';
import { DjangoProject } from './django-project';
import { DockerPublishJobProps } from './docker';
import { ReactCheckJobProps } from './react';
import { ReactProject } from './react-project';


/**
 * Props to configure the Application stack
 */
export interface ApplicationStackProps {
  /**
   * Path to the django project.
   * @default "backend"
   */
  backendPath?: string;

  /**
   * Path to the react project.
   * @default "frontend"
   */
  frontendPath?: string;

  /**
   * If true, run integration tests using docker-compose.
   * @default false
   */
  integrationTests?: boolean;

  /**
   * Base name of the docker images to publish.
   * Will publish $base-frontend and $base-backend images.
   */
  dockerImageBaseName: string;

  /**
   * Name of the django project.
   */
  djangoProjectName: string;

  /**
   * Optional props to pass to the django check job.
   */
  djangoCheckProps?: Partial<DjangoCheckJobProps>;

  /**
   * Optional overrides for the django check job.
   */
  djangoCheckOverrides?: Partial<JobProps>;

  /**
   * Optional props to pass to the django docker publish job.
   */
  djangoDockerProps?: Partial<DockerPublishJobProps>;

  /**
   * Optional overrides for the django docker publish job.
   */
  djangoDockerOverrides?: Partial<JobProps>;

  /**
   * Optional props to pass to the react check job.
   */
  reactCheckProps?: Partial<ReactCheckJobProps>;

  /**
   * Optional overrides for the react check job.
   */
  reactCheckOverrides?: Partial<JobProps>;

  /**
   * Optional props to pass to the react docker publish job.
   */
  reactDockerProps?: Partial<DockerPublishJobProps>;

  /**
   * Optional overrides for the react docker publish job.
   */
  reactDockerOverrides?: Partial<JobProps>;

  /**
   * Optional overrides for the integration tests job.
   */
  integrationOverrides?: Partial<JobProps>;

  /**
   * Optional props to pass to the deploy job.
   */
  deployProps?: Partial<DeployJobProps>;

  /**
   * Optional overrides for the deploy job.
   */
  deployOverrides?: Partial<JobProps>;
}

/**
 * A stack to test, build, and deploy a fullstack application (Django + React).
 */
export class ApplicationStack extends Stack {
  /**
   *
   * @param scope cdkactions App instance.
   * @param config Configuration for the Application stack.
   * @param overrides Optional overrides for the workflow.
   */
  public constructor(scope: Construct, config: ApplicationStackProps, overrides?: Partial<WorkflowProps>) {
    // Build config
    const fullConfig: Required<ApplicationStackProps> = {
      backendPath: 'backend',
      frontendPath: 'frontend',
      integrationTests: false,
      djangoCheckProps: {},
      djangoCheckOverrides: {},
      djangoDockerProps: {},
      djangoDockerOverrides: {},
      reactCheckProps: {},
      reactCheckOverrides: {},
      reactDockerProps: {},
      reactDockerOverrides: {},
      integrationOverrides: {},
      deployProps: {},
      deployOverrides: {},
      ...config,
    };
    // Create stack
    super(scope, 'application');
    const workflow = new Workflow(this, 'build-and-deploy', {
      name: 'Build and Deploy',
      on: 'push',
      ...overrides,
    });

    // Django
    const djangoProject = new DjangoProject(workflow,
      {
        projectName: fullConfig.djangoProjectName,
        path: fullConfig.backendPath,
        imageName: `${fullConfig.dockerImageBaseName}-backend`,
        checkProps: fullConfig.djangoCheckProps,
        checkOverrides: fullConfig.djangoCheckOverrides,
        publishProps: fullConfig.djangoDockerProps,
        publishOverrides: fullConfig.djangoDockerOverrides,
      });

    // React
    const reactProject = new ReactProject(workflow,
      {
        path: fullConfig.frontendPath,
        imageName: `${fullConfig.dockerImageBaseName}-frontend`,
        checkProps: fullConfig.reactCheckProps,
        checkOverrides: fullConfig.reactCheckOverrides,
        publishProps: fullConfig.reactDockerProps,
        publishOverrides: fullConfig.reactDockerOverrides,
      },
    );

    const integrationTestsId = 'integration-tests';
    if (fullConfig.integrationTests) {
      // TODO: finish this
      new CheckoutJob(workflow, integrationTestsId, {
        name: 'Integration Tests',
        runsOn: 'ubuntu-latest',
        needs: [djangoProject.publishJobId, reactProject.publishJobId],
        steps: [{
          name: 'Run docker compose',
          run: 'docker-compose up -d -f docker-compose.test.yaml',
        },
        {
          name: 'Populate backend',
          run: 'docker run backend python manage.py populate',
        },
        {
          name: 'Run integration tests',
          run: 'docker run frontend yarn integration',
        }],
        env: {
          GIT_SHA: '${{ github.sha }}',
        },
        ...fullConfig.integrationOverrides,
      });
    }

    // Deploy
    const deployNeeds = fullConfig.integrationTests ? [integrationTestsId] : [djangoProject.publishJobId, reactProject.publishJobId];
    new DeployJob(workflow, fullConfig.deployProps,
      {
        needs: deployNeeds,
        ...fullConfig.deployOverrides,
      });
  }
}
