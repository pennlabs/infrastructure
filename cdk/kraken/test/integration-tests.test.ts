import { Workflow } from 'cdkactions';
import { IntegrationTestsJob } from '../src';


test('default', () => {
  const workflow = new Workflow(undefined as any, 'workflow', {
    name: 'Workflow',
    on: 'push',
  });
  new IntegrationTestsJob(workflow, {
    testCommand: 'exit 0',
    dockerBuildIds: ['id'],
    dockerImages: ['image'],
  });
  expect(workflow.toGHAction()).toMatchSnapshot();
});

test('no post image publish', () => {
  const workflow = new Workflow(undefined as any, 'workflow', {
    name: 'Workflow',
    on: 'push',
  });
  new IntegrationTestsJob(workflow, {
    testCommand: 'exit 0',
    dockerBuildIds: ['id'],
    dockerImages: ['image'],
    createPostIntegrationPublishJob: false,
  });
  expect(workflow.toGHAction()).toMatchSnapshot();
});
