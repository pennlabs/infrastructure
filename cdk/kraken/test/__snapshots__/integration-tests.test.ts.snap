// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`default 1`] = `
Object {
  "jobs": Object {
    "integration-tests": Object {
      "env": Object {
        "GIT_SHA": "\${{ github.sha }}",
      },
      "name": "Integration Tests",
      "runs-on": "ubuntu-latest",
      "steps": Array [
        Object {
          "uses": "actions/checkout@v2",
        },
        Object {
          "uses": "actions/download-artifact@v2",
        },
        Object {
          "name": "Load docker images",
          "run": "docker load --input id/image.tar",
        },
        Object {
          "name": "Run docker compose",
          "run": "mkdir -p /tmp/test-results
docker-compose -f docker-compose.test.yaml up -d",
        },
        Object {
          "name": "Wait for backend",
          "run": "for try in {1..20}; do
  docker-compose -f docker-compose.test.yaml exec -T backend python manage.py migrate --check && break
  sleep 5
done",
        },
        Object {
          "name": "Populate backend",
          "run": "docker-compose -f docker-compose.test.yaml exec -T backend python manage.py populate",
        },
        Object {
          "name": "Run integration tests",
          "run": "exit 0",
        },
        Object {
          "if": "failure() || github.ref != 'refs/heads/master'",
          "name": "Delete artifacts when no longer needed",
          "uses": "geekyeggo/delete-artifact@v1",
          "with": Object {
            "name": "id",
          },
        },
        Object {
          "if": "failure()",
          "name": "Print logs on failure",
          "run": "docker-compose -f docker-compose.test.yaml logs",
        },
        Object {
          "if": "failure()",
          "name": "Upload artifacts on failure",
          "uses": "actions/upload-artifact@v2",
          "with": Object {
            "name": "cypress-output",
            "path": "/tmp/test-results",
          },
        },
      ],
    },
    "post-integration-publish": Object {
      "if": "github.ref == 'refs/heads/master'",
      "name": "Publish Images",
      "needs": "integration-tests",
      "runs-on": "ubuntu-latest",
      "steps": Array [
        Object {
          "uses": "actions/checkout@v2",
        },
        Object {
          "uses": "actions/download-artifact@v2",
        },
        Object {
          "uses": "geekyeggo/delete-artifact@v1",
          "with": Object {
            "name": "id",
          },
        },
        Object {
          "name": "Load docker images",
          "run": "docker load --input id/image.tar",
        },
        Object {
          "uses": "docker/login-action@v1",
          "with": Object {
            "password": "\${{ secrets.DOCKER_PASSWORD }}",
            "username": "\${{ secrets.DOCKER_USERNAME }}",
          },
        },
        Object {
          "name": "Push docker images",
          "run": "docker push -a image",
        },
      ],
    },
  },
  "name": "Workflow",
  "on": "push",
}
`;

exports[`no post image publish 1`] = `
Object {
  "jobs": Object {
    "integration-tests": Object {
      "env": Object {
        "GIT_SHA": "\${{ github.sha }}",
      },
      "name": "Integration Tests",
      "runs-on": "ubuntu-latest",
      "steps": Array [
        Object {
          "uses": "actions/checkout@v2",
        },
        Object {
          "uses": "actions/download-artifact@v2",
        },
        Object {
          "name": "Load docker images",
          "run": "docker load --input id/image.tar",
        },
        Object {
          "name": "Run docker compose",
          "run": "mkdir -p /tmp/test-results
docker-compose -f docker-compose.test.yaml up -d",
        },
        Object {
          "name": "Wait for backend",
          "run": "for try in {1..20}; do
  docker-compose -f docker-compose.test.yaml exec -T backend python manage.py migrate --check && break
  sleep 5
done",
        },
        Object {
          "name": "Populate backend",
          "run": "docker-compose -f docker-compose.test.yaml exec -T backend python manage.py populate",
        },
        Object {
          "name": "Run integration tests",
          "run": "exit 0",
        },
        Object {
          "if": "failure() || github.ref != 'refs/heads/master'",
          "name": "Delete artifacts when no longer needed",
          "uses": "geekyeggo/delete-artifact@v1",
          "with": Object {
            "name": "id",
          },
        },
        Object {
          "if": "failure()",
          "name": "Print logs on failure",
          "run": "docker-compose -f docker-compose.test.yaml logs",
        },
        Object {
          "if": "failure()",
          "name": "Upload artifacts on failure",
          "uses": "actions/upload-artifact@v2",
          "with": Object {
            "name": "cypress-output",
            "path": "/tmp/test-results",
          },
        },
      ],
    },
  },
  "name": "Workflow",
  "on": "push",
}
`;
